backend/pkg/k8cache/k8cache.go

```diff
@@ -0,0 +1,639 @@
+	r *http.Request,
+) (bool, error) {
+	token := r.Header.Get("Authorization")
+
+	clientset, err := getClientSet(k, token)
+	if err != nil {
+		return false, err
+	}
+
+	last, kubeVerb := GetKindAndVerb(r)
+
+	review := &authorizationv1.SelfSubjectAccessReview{
+		Spec: authorizationv1.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationv1.ResourceAttributes{
+				Resource: last,
+				Verb:     kubeVerb,
+			},
+		},
+	}
+
+	result, err := clientset.AuthorizationV1().SelfSubjectAccessReviews().Create(
+		context.TODO(),
+		review,
+		metav1.CreateOptions{},
+	)
+
+	return result.Status.Allowed, err
+}
+
+// LoadFromCache ensure that if the user has the permission to view the resources then
+// it will check if the generated key is found in the cache if the key is present in the cache
+// then it will return directly to the client in []byte form and returns true ,Otherwise it will return false.
+func LoadFromCache(k8scache cache.Cache[string], isAllowed bool,
+	key string, w http.ResponseWriter, r *http.Request,
+) (bool, error) {
+	k8Resource, err := k8scache.Get(context.Background(), key)
+	if err == nil && strings.TrimSpace(k8Resource) != "" && isAllowed {
+		var cachedData CachedResponseData
+
+		cachedData, err := UnmarshalCacheData(k8Resource, cachedData)
+		if err != nil {
+			return false, err
+		}
+
+		SetHeader(cachedData, w)
+
+		_, writeErr := w.Write([]byte(cachedData.Body))
+		if writeErr == nil {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// RequestK8ClusterAPIAndStore ensures if the key was not found inside the cache then this will make actual call to k8's
+// and this will capture the response body and convert the captured response to string.
+// After converting it will store the response with the key and TTL of 10*min.
+func RequestK8ClusterAPIAndStore(k8scache cache.Cache[string],
+	url *url.URL,
+	rcw *responseCapture,
+	r *http.Request,
+	key string,
+) error {
+	capturedHeaders := rcw.Header()
+	encoding := capturedHeaders.Get("Content-Encoding")
+	bodyBytes := rcw.Body.Bytes()
+
+	dcmpBody, err := GetResponseBody(bodyBytes, encoding)
+	if err != nil {
+		return err
+	}
+
+	headersToCache := FilterHeaderForCache(capturedHeaders, encoding)
+
+	if !strings.Contains(url.Path, "selfsubjectrulesreviews") {
+		cachedData := CachedResponseData{
+			StatusCode: rcw.StatusCode,
+			Headers:    headersToCache,
+			Body:       dcmpBody,
+		}
+
+		jsonBytes, err := MarshalToStore(cachedData)
+		if err != nil {
+			return err
+		}
+
+		if !strings.Contains(string(jsonBytes), "Failure") {
+			if err = k8scache.SetWithTTL(context.Background(), key, string(jsonBytes), 10*time.Minute); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// StoreAfterAuthError Stores resource(pods , nodes , etc) and returns to client
+// if we get error while Authorizing user's permissions for every resources.
+func StoreAfterAuthError(k8scache cache.Cache[string], isAllowed bool, next http.Handler, key string,
```

Comment:
This function is named `StoreAfterAuthError` based on where this function is called in the cachemiddleware. It should be names something like `ServeFromCacheOrForwardToK8s` based on what action this function performs.